### PR TITLE
TRISO bug fix and unit tests

### PR DESCRIPTION
**Pull request by amandalund**
This PR fixes an issue in the close random packing algorithm where TRISO particles packed in a cylinder or sphere could be located outside the domain. It also introduces some unit tests for TRISO modeling.

### CODE REVIEW

File: tests/unit_tests/test_model_triso.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                         'radius': 0.,
+                         'volume': 0.75**3},
+                        {'shape': 'cylinder',
+                         'length': 0.5,
+                         'radius': 0.5,
+                         'volume': 0.5*pi*0.5**2},
+                        {'shape': 'sphere',
+                         'length': 0.,
+                         'radius': 0.5,
+                         'volume': 4/3*pi*0.5**3}])
```
Comment (paulromano) [inline tests/unit_tests/test_model_triso.py]: Seems a little awkward to have a decorator that spans so many lines. Maybe change to something like:
```Python
domain_params = [
    {'shape': 'cube', 'length': 0.75, 'radius': 0., 'volume': 0.75**3},
    {'shape': 'cylinder', 'length': 0.5, 'radius': 0.5, 'volume': 0.5*pi*0.5**2},
    {'shape': 'sphere', 'length': 0., 'radius': 0.5, 'volume': 4/3*pi*0.5**3}
]


@pytest.fixture(scope='module', params=domain_params,
                ids=['cube', 'cylinder', 'sphere'])
```

File: tests/unit_tests/test_model_triso.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    centers = [t.center for t in trisos]
+
+    # Create KD tree for quick nearest neighbor search
+    tree = scipy.spatial.cKDTree(centers)
+
+    # Find distance to nearest neighbor for all particles
+    d = tree.query(centers, k=2)[0]
+
+    # Get the smallest distance between any two particles
+    d_min = min(d[:,1])
```
Comment (paulromano) [inline tests/unit_tests/test_model_triso.py]: PEP8: `min(d[:, 1])`

**Comment by amandalund** [comment]
In the CRP algorithm, overlaps between particles are eliminated in each iteration both directly by moving the closest pair of particles apart and indirectly by shrinking the particle radius. The step of separating the particles can cause them to cross the domain boundary -- if this happens, they are shifted back inside the domain perpendicular to the surface. This PR doesn't change this method of handling particles near the boundary, it just fixes an oversight on my part in extending the implementation to all domain shapes.

To make sure this treatment of particles near the surface leads to a realistic configuration of randomly packed spheres, I looked at the porosity as a function of the radial distance from the cylindrical surface. There have been a bunch of experimental studies on the distribution of close random packings of spheres in rigid containers (e.g. [Benenati, R. F. and Brosilow, C. B., Void Fraction Distribution in Beds of Spheres, AIChE J., 8, 359-361 (1962)](http://www.academia.edu/16629663/Void_fraction_distribution_in_beds_of_spheres); [Goodling, J. S., Vachon, R. I., Stelpflung, W. S., Ying, S. J. and Khader, M. S., Radial Porosity Distribution in Cylindrical Beds Packed with Spheres, Powder Technol., 35, 23-29 (1983)](https://www.sciencedirect.com/science/article/pii/0032591083850220)). They have shown that these packings exhibit a wall effect: spheres near the surface have a more ordered structure since they come in contact with it and conform to its shape, and the order disappears away from the wall. A plot of porosity as a function of radial distance looks like a damped oscillator, approaching unity at the wall (since the spheres only make contact with the surface at a single point), reaching a minimum one sphere radius from the surface, and then oscillating with decreasing amplitude and approaching the mean porosity of a random packing towards the interior.

The plot below shows the porosity for a simulated close random packing of spheres in a cylinder, where the diameter of the cylinder is equal to 20 sphere diameters and the packing fraction is 0.6. The simulation looks similar to the experimental results; the porosity is maximum at the wall and minimum one sphere radius from it, and the damped oscillations approach the expected equilibrium away from the wall.

 porosity

**Comment by paulromano** [comment]
Thanks @amandalund!

**Pull request merged**